rmw_zenoh_cpp/src/rmw_service.cpp

```diff
@@ -0,0 +1,419 @@
+    service_members->response_members_->data
+  );
+
+  service_data->typesupport_identifier_ = type_support->typesupport_identifier;
+  service_data->request_type_support_impl_ = request_members;
+  service_data->response_type_support_impl_ = response_members;
+
+  // Allocate and in-place assign new typesupport instances
+  service_data->request_type_support_ = static_cast<RequestTypeSupport_cpp *>(
+    allocator->allocate(sizeof(RequestTypeSupport_cpp *), allocator->state)
+  );
+  new(service_data->request_type_support_) RequestTypeSupport_cpp(service_members);
+  if (!service_data->request_type_support_) {
+    RMW_SET_ERROR_MSG("failed to allocate RequestTypeSupport");
+    allocator->deallocate(service_data->request_type_support_, allocator->state);
+    allocator->deallocate(service->data, allocator->state);
+    allocator->deallocate(service, allocator->state);
+    return nullptr;
+  }
+
+  service_data->response_type_support_ = static_cast<ResponseTypeSupport_cpp *>(
+    allocator->allocate(sizeof(ResponseTypeSupport_cpp *), allocator->state)
+  );
+  new(service_data->response_type_support_) ResponseTypeSupport_cpp(service_members);
+  if (!service_data->response_type_support_) {
+    RMW_SET_ERROR_MSG("failed to allocate ResponseTypeSupport");
+    allocator->deallocate(service_data->request_type_support_, allocator->state);
+    allocator->deallocate(service_data->response_type_support_, allocator->state);
+    allocator->deallocate(service->data, allocator->state);
+    allocator->deallocate(service, allocator->state);
+    return nullptr;
+  }
+
+  // // Assign node pointer
+  service_data->node_ = node;
+
+  // Init Zenoh subscriber for request messages
+  service_data->zn_request_subscriber_ = zn_declare_subscriber(
+    service_data->zn_session_,
+    service_data->zn_request_topic_key_,
+    zn_subinfo_default(),  // NOTE(CH3): Default for now
+    service_data->zn_request_sub_callback
+  );
+
+  // Init Zenoh queryable for availability checking
+  service_data->zn_queryable_ = zn_declare_queryable(
+    s, service->service_name, EVAL, [](ZNQuery * query){
+      const zn_string * resource = zn_query_res_name(query);
+      const zn_string * predicate = zn_query_predicate(query);
+
+      std::string key(resource->val, resource->len);
+      std::string response("available");  // NOTE(CH3): The contents actually don't matter...
+
+      zn_send_reply(query, key.c_str(), (const unsigned char *)response.c_str(), response.length());
+    }
+  );
+  if (service_data->zn_queryable_ == 0) {
+    RMW_SET_ERROR_MSG("failed to create availability queryable for service");
+
+    zn_undeclare_subscriber(service_data->zn_request_subscriber_);
+
+    allocator->deallocate(service_data->request_type_support_, allocator->state);
+    allocator->deallocate(service_data->response_type_support_, allocator->state);
+    allocator->deallocate(service->data, allocator->state);
+    allocator->deallocate(service, allocator->state);
```

Comment:
Does this leak `service->service_name`?

Reply:
Yup!

Updated: 23497d4430f890dc62fb9995c66fb2be8602b51e

(Sorry for the massive commit, it's kind of intertwined...)

Reply:
Yus

Updated: 23497d4

(Sorry for the massive commit, it's kind of intertwined...)